src/ServiceControl/Monitoring/EndpointInstanceMonitor.cs

```diff
@@ -0,0 +1,122 @@
+namespace ServiceControl.Monitoring
+{
+    using System;
+    using ServiceControl.CompositeViews.Endpoints;
+    using ServiceControl.Contracts.HeartbeatMonitoring;
+    using ServiceControl.Contracts.Operations;
+    using ServiceControl.EndpointControl.Contracts;
+    using ServiceControl.HeartbeatMonitoring;
+    using ServiceControl.Infrastructure.DomainEvents;
+
+    class EndpointInstanceMonitor
+    {
+        public EndpointInstanceId Id { get; }
+        private DateTime? lastSeen;
+        private HeartbeatStatus status;
+
+        public bool Monitored { get; private set; }
+
+        public EndpointInstanceMonitor(EndpointInstanceId endpointInstanceId, bool monitored)
+        {
+            Id = endpointInstanceId;
+            Monitored = monitored;
+        }
+
+        public void EnableMonitoring()
+        {
+            DomainEvents.Raise(new MonitoringEnabledForEndpoint { Endpoint = Convert(Id) });
+            Monitored = true;
+        }
+
+        public void DisableMonitoring()
+        {
+            DomainEvents.Raise(new MonitoringDisabledForEndpoint { Endpoint = Convert(Id) });
+            Monitored = false;
+        }
+
+        public void UpdateStatus(HeartbeatStatus newStatus, DateTime? latestTimestamp)
+        {
+            if (newStatus != status)
+            {
+                RaiseStateChangeEvents(newStatus, latestTimestamp);
+            }
+
+            lastSeen = latestTimestamp;
+            status = newStatus;
+        }
+
+        private void RaiseStateChangeEvents(HeartbeatStatus newStatus, DateTime? latestTimestamp)
+        {
+            if (newStatus == HeartbeatStatus.Alive)
+            {
+                if (status == HeartbeatStatus.Unknown)
+                {
+                    // NOTE: If an endpoint starts randomly sending heartbeats we monitor it by default
+                    // NOTE: This means we'll start monitoring endpoints sending heartbeats after a restart
+                    Monitored = true;
+                    DomainEvents.Raise(new HeartbeatingEndpointDetected
```

Comment:
So after a restart all endpoints will be monitored again once they start sending messages? Or are the "don't monitor" flags persisted ?